retrofit2/src/main/java/com/linecorp/armeria/client/retrofit2/ArmeriaCallFactory.java

```diff
@@ -196,7 +186,18 @@ public Response execute() throws IOException {
 
         @Override
         public void enqueue(Callback callback) {
-            createRequest().subscribe(callFactory.subscriberFactory.create(this, callback, request));
+            if (httpResponse != null) {
+                throw new IllegalStateException("executed already");
+            }
+
+            // Run atomically
+            if (executionStateUpdater.compareAndSet(this, ExecutionState.IDLE, ExecutionState.RUNNING)) {
```

Comment:
I tried to make the use of CAS instead of using Reentrant Lock.
So, when the compare failed, `enqueue` returns with failure callback.